### ISSUE TITLE
Support Python 3.11

### ISSUE THREAD

**Issue by EwoutH** [pytorch/pytorch#86566]
Currently Python 3.11 isn't fully supported in Pytorch yet, including testing in CI and publishing wheels to PyPI.

Python 3.11 is currently in Release Candidate state and will be released in about two weeks. Major packages like NumPy, Pandas, SciPy and Matplotlib have already uploaded their Python 3.11 wheels to PyPI. 

Python 3.11 new features include:

> - [PEP 657](https://www.python.org/dev/peps/pep-0657/) -- Include Fine-Grained Error Locations in Tracebacks
> - [PEP 654](https://www.python.org/dev/peps/pep-0654/) -- Exception Groups and except*
> - [PEP 673](https://www.python.org/dev/peps/pep-0673/) -- Self Type
> - [PEP 646](https://www.python.org/dev/peps/pep-0646/)-- Variadic Generics
> - [PEP 680](https://www.python.org/dev/peps/pep-0680/)-- tomllib: Support for Parsing TOML in the Standard Library
> - [PEP 675](https://www.python.org/dev/peps/pep-0675/)-- Arbitrary Literal String Type
> - [PEP 655](https://www.python.org/dev/peps/pep-0655/)-- Marking individual TypedDict items as required or potentially-missing
> - [bpo-46752](https://bugs.python.org/issue46752)-- Introduce task groups to asyncio
> - The [Faster Cpython Project](https://github.com/faster-cpython) is already yielding some exciting results: this version of CPython 3.11 is ~ 19% faster on the geometric mean of the [PyPerformance benchmarks](https://speed.python.org/), compared to 3.10.0.

Pytorch is an import package in the deep learning stack, so support at release will help speed up Python 3.11 adoption.

It would be great to have full Python 3.11 support in Pytorch by the time Python 3.11.0 gets released, ([expected](https://peps.python.org/pep-0664/) Monday October 6th, 2022). This includes testing in CI and publishing wheels to PyPI.

This issue can be used as a tracking issue. This includes:

- [x] Pytorch builds fully on Python 3.11
- [x] All Pytorch tests pass on Python 3.11
- [x] All CI is run and green on Python 3.11
- [x] Wheels are uploaded to PyPI for at least one Pytorch release
  - [x] Manylinux (x86_64)
  - [x] Manylinux (aarch64)
  - [x] Windows
  - [x] macOS

cc @seemethere @malfet @pytorch/pytorch-dev-infra @albanD

**Comment by malfet** [comment]
Nightly builds of python-3.11 wheels is already published to https://download.pytorch.org/whl/nightly but CI can be tricky, in particular around JIT support (which probably needs a separate task)

**Comment by albanD** [comment]
I feel like the main blocker for enabling more support is that (at the time where this was added at least) conda didn't had 3.11 support.
This meant that all our usual test environment couldn't be upgraded easily.

**Comment by malfet** [comment]
@dagitses suggests that it's probably OK to disable JIT tests and have a followup issue for supporting JIT in 3.11

**Comment by albanD** [comment]
cc @davidberard98 there will most likely be some non-trivial work needed for JIT to support this that should be planned for ahead of time.

**Comment by rijobro** [comment]
conda now has 3.11 support if that is still a blocker! `conda install -c conda-forge python`

**Comment by johnnynunez** [comment]
Yeah! Python 3.11 support please. It should be the default.
Python 3.11 is between 10-60% faster than Python 3.10. On average, we measured a 1.25x speedup on the standard benchmark suite. See Faster CPython for details.
https://docs.python.org/3.11/whatsnew/3.11.html
https://www.phoronix.com/news/Python-3.11-Released

**Comment by weiwangmeta** [comment]
> Yeah! Python 3.11 support please. It should be the default.

cc @malfet @atalman

**Comment by seemethere** [comment]
> conda now has 3.11 support if that is still a blocker! `conda install -c conda-forge python`

conda-forge introduces a lot of instability to our CI pipelines so adding conda support should be gated on the default channels having python 3.11 support not conda-forge

**Comment by johnnynunez** [comment]
Benchmark https://www.phoronix.com/review/python-311-performance

**Comment by MilesCranmer** [comment]
What is the status of this? Python 3.11 was released ~two weeks ago now so it would be great to have torch wheels available relatively soon for downstream projects to work with.

I would be happy nightly versions too if a stable release will take some more time, although the current binaries are exclusively linux x86: https://download.pytorch.org/whl/nightly/torch/.

**Comment by Birch-san** [comment]
> Nightly builds of python-3.11 wheels is already published to https://download.pytorch.org/whl/nightly

how do I install these?

https://conda.anaconda.org/pytorch-nightly/osx-arm64  
<img width="707" alt="image" src="https://user-images.githubusercontent.com/6141784/200122962-a4d99a1a-03e9-4265-ac8b-2a66ecd79700.png">

I only see python 3.10 builds there.

conda couldn't find any 3.11-compatible pytorch build in the `pytorch-nightly` channel.

```
conda create -n 11diffnightly -c pytorch-nightly -c defaults -c conda-forge python==3.11.0 pytorch
Collecting package metadata (current_repodata.json): done
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: | 
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
Examining pytorch:
^failed

UnsatisfiableError: The following specifications were found to be incompatible with each other:

Output in format: Requested package -> Available versions

Package python conflicts for:
python==3.11.0
pytorch -> python[version='>=3.10,<3.11.0a0|>=3.10,<3.11.0a0|>=3.8,<3.9.0a0|>=3.9,<3.10.0a0|>=3.8,<3.9.0a0|>=3.9,<3.10.0a0',build=*_cpython]
pytorch -> typing_extensions -> python[version='3.10.*|3.6.12|3.6.12|3.7.10|3.7.10|3.7.12|3.8.12|3.9.10|>=3.11,<3.12.0a0|>=3.11,<3.12.0a0|>=3.5|>=3.6|>=3.7|3.7.9|3.6.9|3.6.9|3.6.9|3.6.9|>=3.8|3.8.*|3.9.*',build='2_73_pypy|4_73_pypy|5_73_pypy|5_73_pypy|0_73_pypy|1_73_pypy|*_cpython|0_73_pypy|0_73_pypy|0_73_pypy|3_73_pypy|1_73_pypy|0_73_pypy']The following specifications were found to be incompatible with your system:

  - feature:/osx-arm64::__osx==12.5=0
  - feature:|@/osx-arm64::__osx==12.5=0
  - pytorch -> __osx[version='>=11.0']

Your installed version is: 12.5
```

**Comment by johnnynunez** [comment]
> > Nightly builds of python-3.11 wheels is already published to https://download.pytorch.org/whl/nightly
> 
> how do I install these?
> 
> https://conda.anaconda.org/pytorch-nightly/osx-arm64 <img alt="image" width="707" src="https://user-images.githubusercontent.com/6141784/200122962-a4d99a1a-03e9-4265-ac8b-2a66ecd79700.png">
> 
> I only see python 3.10 builds there.
> 
> conda couldn't find any 3.11-compatible pytorch build in the `pytorch-nightly` channel:
> 
> ```
> conda create -n 11diffnightly -c pytorch-nightly -c defaults -c conda-forge python==3.11.0 pytorch
> Collecting package metadata (current_repodata.json): done
> Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
> Collecting package metadata (repodata.json): done
> Solving environment: | 
> Found conflicts! Looking for incompatible packages.
> This can take several minutes.  Press CTRL-C to abort.
> Examining pytorch:
> ^failed
> 
> UnsatisfiableError: The following specifications were found to be incompatible with each other:
> 
> Output in format: Requested package -> Available versions
> 
> Package python conflicts for:
> python==3.11.0
> pytorch -> python[version='>=3.10,<3.11.0a0|>=3.10,<3.11.0a0|>=3.8,<3.9.0a0|>=3.9,<3.10.0a0|>=3.8,<3.9.0a0|>=3.9,<3.10.0a0',build=*_cpython]
> pytorch -> typing_extensions -> python[version='3.10.*|3.6.12|3.6.12|3.7.10|3.7.10|3.7.12|3.8.12|3.9.10|>=3.11,<3.12.0a0|>=3.11,<3.12.0a0|>=3.5|>=3.6|>=3.7|3.7.9|3.6.9|3.6.9|3.6.9|3.6.9|>=3.8|3.8.*|3.9.*',build='2_73_pypy|4_73_pypy|5_73_pypy|5_73_pypy|0_73_pypy|1_73_pypy|*_cpython|0_73_pypy|0_73_pypy|0_73_pypy|3_73_pypy|1_73_pypy|0_73_pypy']The following specifications were found to be incompatible with your system:
> 
>   - feature:/osx-arm64::__osx==12.5=0
>   - feature:|@/osx-arm64::__osx==12.5=0
>   - pytorch -> __osx[version='>=11.0']
> 
> Your installed version is: 12.5
> ```

<img width="997" alt="Screenshot 2022-11-05 at 14 50 17" src="https://user-images.githubusercontent.com/22727137/200123165-11b2cc14-4e6f-483a-95be-de55577ecdff.png">

https://pytorch.org/blog/PyTorch-1.13-release/

**Comment by Birch-san** [comment]
okay, so Linux x86 only? thanks for the link.  
if an osx-arm64 build were made available: I'd happily try it out.

**Comment by afvca** [comment]
Will Python 3.11 support get into the [1.13.1 milestone](https://github.com/pytorch/pytorch/milestone/34)?

**Comment by seemethere** [comment]
> Will Python 3.11 support get into the [1.13.1 milestone](https://github.com/pytorch/pytorch/milestone/34)?

No unfortunately not, we no longer add support for new platforms (like python) during patch releases

**Comment by Liso7150** [comment]
I wonder if someone updates it in any time, but is very needed dependency, for us all, and 3.11 is futher more fast than 3.10

**Comment by strngtn** [comment]
> I wonder if someone updates it in any time, but is very needed dependency, for us all, and 3.11 is futher more fast than 3.10

It is really bad this issue has been opened for two months but no progress.

**Comment by donhuvy** [comment]
@malfet you said
```python
Nightly builds of python-3.11 wheels is already published to https://download.pytorch.org/whl/nightly but CI can be tricky, in particular around JIT support (which probably needs a separate task)
```

but in my PC

```python
Microsoft Windows [Version 10.0.22621.963]
(c) Microsoft Corporation. All rights reserved.

C:\Users\donhu>pip3 install torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/cu117
Looking in indexes: https://pypi.org/simple, https://download.pytorch.org/whl/cu117
ERROR: Could not find a version that satisfies the requirement torch (from versions: none)ERROR: No matching distribution found for torch

C:\Users\donhu>pip3 install --pre torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/nightly/cu117
Looking in indexes: https://pypi.org/simple, https://download.pytorch.org/whl/nightly/cu117
ERROR: Could not find a version that satisfies the requirement torch (from versions: none)
ERROR: No matching distribution found for torch

C:\Users\donhu>python --version
Python 3.11.1

C:\Users\donhu>
```

**Comment by jsonmona** [comment]
@donhuvy

It looks like cp311 (Python 3.11) release wheel for linux is already on stable. But other platforms like windows or mac is still not supported.

**Comment by donhuvy** [comment]
We should add information about this fact of nightly version: Not run on Windows 11 x64 or Windows OS x64. It will make us avoid lost time for confusting or questions or looking for solution.

**Comment by EwoutH** [comment]
I updated the starting post to reflect that Linux x64 wheels are available for Python 3.11, but others are not.

I see there is now also a [Python 3.11](https://github.com/pytorch/pytorch/milestone/35) milestone. Hopefully this helps keeping track of 3.11 progress.

To the Pytorch team: It would be hugely appreciated if 3.11 support could be prioritized, and if fully support for all platforms is in the Pytorch 2.0 release.

**Comment by payne911** [comment]
@donhuvy as a Java dev trying to install pytorch with the latest Python version available, I was also surprised to see the lack of retrocompatibility within the Python ecosystem.
I was aware of the break between Python 2 and Python 3, but it seems like even minor version bumps may lead to compatibility issue. What a let down. :(

**Comment by donhuvy** [comment]
@payne911 , My advice: use old version of Python for system. For example, Google Colab pro use Python 3.8 at this time. Don't try latest version in Python eco-system. You should use Python 3.8 , 3.9  for easy and avoid error.

**Comment by malfet** [comment]
PyTorch nightly wheels should be available for all OSes right now, but conda package is still pending.

**Comment by johnnynunez** [comment]
> PyTorch nightly wheels should be available for all OSes right now, but conda package is still pending.

thanks you so much!!

**Comment by albanD** [comment]
@payne911 this is kind of off topic for this but I want to mention that the number of backward compatibility breaking changes in the python code itself are very low in practice. The two reasons this takes time are that it is a lot of Infra work (and in particular getting all dependencies running, including conda having a 3.11 build) and that we use bleeding edge CPython features (tightly integrated with their internal representation and C API, at this level backward compatibility is a lot more challenging and some amount of breakage is expected for the language to be able to move forward. But the number of projects impacted by these changes is really minor compared to the overal ecosystem).

**Comment by payne911** [comment]
> PyTorch nightly wheels should be available for all OSes right now, but conda package is still pending.

Is it normal that I get this, then?

![image](https://user-images.githubusercontent.com/38117856/216666778-39afc218-158d-4c53-b33c-1faff64f3892.png)

Coming from 
![image](https://user-images.githubusercontent.com/38117856/216666872-d18a77bb-609f-4ea9-8f0b-e52c90838196.png)

**Comment by albanD** [comment]
@payne911 yes this is expected. Domain libraries don't have binaries yet.
You can remove `torchvision torchaudio` from the command to make this work.

**Comment by INVISIBLE-JUMPROPE** [comment]
Hi, I'm not the most technical programmer, but I've been trying to install TA_lib and Numpy for a while now without success. Python --version returns 3.11.1   and pip is fully up to date. Am I doing something wrong?

**Comment by johnnynunez** [comment]
> Hi, I'm not the most technical programmer, but I've been trying to install TA_lib and Numpy for a while now without success. Python --version returns 3.11.1 and pip is fully up to date. Am I doing something wrong?

you have to install miniconda or python binaries with 3.10 or 3.9 version.
After this install pytorch version(nightly is beta version)

**Comment by INVISIBLE-JUMPROPE** [comment]
Is there a proper way for me to roll-back my installation? Or use anaconda
to create an environment with 3.9 / 3.10? And then miniconda/ python
binaries>then pytorch?
Thanks

**Comment by donhuvy** [comment]
![image](https://user-images.githubusercontent.com/1328316/217039476-8c3f39c3-fb8f-48fb-ad58-ac37cdd9af73.png)

```
Microsoft Windows [Version 10.0.22621.1105]
(c) Microsoft Corporation. All rights reserved.

C:\Windows\System32>pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu118
WARNING: pip is configured with locations that require TLS/SSL, however the ssl module in Python is not available.
Looking in indexes: https://download.pytorch.org/whl/nightly/cu118, https://pypi.ngc.nvidia.com
WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /whl/nightly/cu118/torch/
WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /whl/nightly/cu118/torch/
WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /whl/nightly/cu118/torch/
WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /whl/nightly/cu118/torch/
WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /whl/nightly/cu118/torch/
Could not fetch URL https://download.pytorch.org/whl/nightly/cu118/torch/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='download.pytorch.org', port=443): Max retries exceeded with url: /whl/nightly/cu118/torch/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.")) - skipping
WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
Could not fetch URL https://pypi.ngc.nvidia.com/torch/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='pypi.ngc.nvidia.com', port=443): Max retries exceeded with url: /torch/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.")) - skipping
ERROR: Could not find a version that satisfies the requirement torch (from versions: none)
ERROR: No matching distribution found for torch
WARNING: pip is configured with locations that require TLS/SSL, however the ssl module in Python is not available.
Could not fetch URL https://download.pytorch.org/whl/nightly/cu118/pip/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='download.pytorch.org', port=443): Max retries exceeded with url: /whl/nightly/cu118/pip/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.")) - skipping
Could not fetch URL https://pypi.ngc.nvidia.com/pip/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='pypi.ngc.nvidia.com', port=443): Max retries exceeded with url: /pip/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.")) - skipping

C:\Windows\System32>
```

How to install success?

**Comment by johnnynunez** [comment]
> ![image](https://user-images.githubusercontent.com/1328316/217039476-8c3f39c3-fb8f-48fb-ad58-ac37cdd9af73.png)
> 
> ```
> Microsoft Windows [Version 10.0.22621.1105]
> (c) Microsoft Corporation. All rights reserved.
> 
> C:\Windows\System32>pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu118
> WARNING: pip is configured with locations that require TLS/SSL, however the ssl module in Python is not available.
> Looking in indexes: https://download.pytorch.org/whl/nightly/cu118, https://pypi.ngc.nvidia.com
> WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /whl/nightly/cu118/torch/
> WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /whl/nightly/cu118/torch/
> WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /whl/nightly/cu118/torch/
> WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /whl/nightly/cu118/torch/
> WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /whl/nightly/cu118/torch/
> Could not fetch URL https://download.pytorch.org/whl/nightly/cu118/torch/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='download.pytorch.org', port=443): Max retries exceeded with url: /whl/nightly/cu118/torch/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.")) - skipping
> WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
> WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
> WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
> WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
> WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
> Could not fetch URL https://pypi.ngc.nvidia.com/torch/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='pypi.ngc.nvidia.com', port=443): Max retries exceeded with url: /torch/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.")) - skipping
> ERROR: Could not find a version that satisfies the requirement torch (from versions: none)
> ERROR: No matching distribution found for torch
> WARNING: pip is configured with locations that require TLS/SSL, however the ssl module in Python is not available.
> Could not fetch URL https://download.pytorch.org/whl/nightly/cu118/pip/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='download.pytorch.org', port=443): Max retries exceeded with url: /whl/nightly/cu118/pip/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.")) - skipping
> Could not fetch URL https://pypi.ngc.nvidia.com/pip/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='pypi.ngc.nvidia.com', port=443): Max retries exceeded with url: /pip/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.")) - skipping
> 
> C:\Windows\System32>
> ```
> 
> How to install success?
only exists torch on 3.11 at this moment

**Comment by juliarscat** [comment]
To install torch on 3.11:
pip install --pre torch -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html

**Comment by wkoot** [comment]
> To install torch on 3.11: pip install --pre torch -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html
Verified working on Windows and within WSL, tnx!

**Comment by TaridaGeorge** [comment]
For me it's not working:

```
pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu117
Looking in indexes: https://download.pytorch.org/whl/nightly/cu117
Collecting torch
  Downloading https://download.pytorch.org/whl/nightly/cu117/torch-2.0.0.dev20230209%2Bcu117-cp311-cp311-linux_x86_64.whl (1837.7 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 GB 2.0 MB/s eta 0:00:00
Collecting torchvision
  Downloading https://download.pytorch.org/whl/nightly/cu117/torchvision-0.15.0.dev20230209%2Bcu117-cp311-cp311-linux_x86_64.whl (6.2 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.2/6.2 MB 8.1 MB/s eta 0:00:00
ERROR: Could not find a version that satisfies the requirement torchaudio (from versions: none)
ERROR: No matching distribution found for torchaudio

[notice] A new release of pip available: 22.3 -> 23.0
[notice] To update, run: pip install --upgrade pip
george@g6:~/Projects/proj1$ python --version
Python 3.11.0
```

**Comment by Somiona** [comment]
Currently pytorch with cuda and cp311 can be installed on windows! To install cuda enabled torch on cu311
`pip3 install --pre torch --index-url https://download.pytorch.org/whl/nightly/cu118`

![image](https://user-images.githubusercontent.com/20739885/218238308-4d5b4afb-88c5-490f-951a-f7b8c591c7f7.png)

![image](https://user-images.githubusercontent.com/20739885/218238336-6b6bf828-5345-415e-a66f-88b61d0210bd.png)

![image](https://user-images.githubusercontent.com/20739885/218238419-5f703235-102e-4c34-add3-46b91bfe8f8c.png)

**Comment by donhuvy** [comment]
I catch error
![Uploading image.png…]()

```python
Microsoft Windows [Version 10.0.22621.1105]
(c) Microsoft Corporation. All rights reserved.

C:\Windows\System32>pip3 install --pre torch --index-url https://download.pytorch.org/whl/nightly/cu118
WARNING: pip is configured with locations that require TLS/SSL, however the ssl module in Python is not available.
Looking in indexes: https://download.pytorch.org/whl/nightly/cu118, https://pypi.ngc.nvidia.com
WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /whl/nightly/cu118/torch/
WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /whl/nightly/cu118/torch/
WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /whl/nightly/cu118/torch/
WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /whl/nightly/cu118/torch/
WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /whl/nightly/cu118/torch/
Could not fetch URL https://download.pytorch.org/whl/nightly/cu118/torch/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='download.pytorch.org', port=443): Max retries exceeded with url: /whl/nightly/cu118/torch/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.")) - skipping
WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
Could not fetch URL https://pypi.ngc.nvidia.com/torch/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='pypi.ngc.nvidia.com', port=443): Max retries exceeded with url: /torch/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.")) - skipping
ERROR: Could not find a version that satisfies the requirement torch (from versions: none)
ERROR: No matching distribution found for torch
WARNING: pip is configured with locations that require TLS/SSL, however the ssl module in Python is not available.
Could not fetch URL https://download.pytorch.org/whl/nightly/cu118/pip/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='download.pytorch.org', port=443): Max retries exceeded with url: /whl/nightly/cu118/pip/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.")) - skipping
Could not fetch URL https://pypi.ngc.nvidia.com/pip/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='pypi.ngc.nvidia.com', port=443): Max retries exceeded with url: /pip/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.")) - skipping

C:\Windows\System32>pip3 install --pre torch --index-url http://download.pytorch.org/whl/nightly/cu118
WARNING: pip is configured with locations that require TLS/SSL, however the ssl module in Python is not available.
Looking in indexes: http://download.pytorch.org/whl/nightly/cu118, https://pypi.ngc.nvidia.com
WARNING: The repository located at download.pytorch.org is not a trusted or secure host and is being ignored. If this repository is available via HTTPS we recommend you use HTTPS instead, otherwise you may silence this warning and allow it anyway with '--trusted-host download.pytorch.org'.
WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /torch/
Could not fetch URL https://pypi.ngc.nvidia.com/torch/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='pypi.ngc.nvidia.com', port=443): Max retries exceeded with url: /torch/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.")) - skipping
ERROR: Could not find a version that satisfies the requirement torch (from versions: none)
ERROR: No matching distribution found for torch
WARNING: pip is configured with locations that require TLS/SSL, however the ssl module in Python is not available.
WARNING: The repository located at download.pytorch.org is not a trusted or secure host and is being ignored. If this repository is available via HTTPS we recommend you use HTTPS instead, otherwise you may silence this warning and allow it anyway with '--trusted-host download.pytorch.org'.
Could not fetch URL https://pypi.ngc.nvidia.com/pip/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='pypi.ngc.nvidia.com', port=443): Max retries exceeded with url: /pip/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.")) - skipping

C:\Windows\System32>
```

**Comment by alex77g2** [comment]
> I catch error ![image](https://user-images.githubusercontent.com/1328316/218241541-764a02dc-cdc2-48a6-b22f-80bb9632054b.png)
> 
> ```python
> Microsoft Windows [Version 10.0.22621.1105]
> (c) Microsoft Corporation. All rights reserved.
> 
> C:\Windows\System32>pip3 install --pre torch --index-url https://download.pytorch.org/whl/nightly/cu118
> WARNING: pip is configured with locations that require TLS/SSL, however the ssl module in Python is not available.
...
> C:\Windows\System32>pip3 install --pre torch --index-url http://download.pytorch.org/whl/nightly/cu118
> WARNING: pip is configured with locations that require TLS/SSL, however the ssl module in Python is not available.
...
> WARNING: pip is configured with locations that require TLS/SSL, however the ssl module in Python is not available.
> WARNING: The repository located at download.pytorch.org is not a trusted or secure host and is being ignored. If this repository is available via HTTPS we recommend you use HTTPS instead, otherwise you may silence this warning and allow it anyway with '--trusted-host download.pytorch.org'.
> Could not fetch URL https://pypi.ngc.nvidia.com/pip/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='pypi.ngc.nvidia.com', port=443): Max retries exceeded with url: /pip/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.")) - skipping
> 
> C:\Windows\System32>
> ```

I think you are missing SSL-module (or perhaps access rights or login missing).
"require TLS/SSL, however the ssl module in Python is not available"
I cannot solve this, but I think thats your issue.

**Comment by donhuvy** [comment]
I hope some one help me. I like practice with PyTorch.

**Comment by devdave** [comment]
@donhuvy One thing to try is to make sure Python is completely in your $PATH.  For my M$ workstation `echo %PATH%` has

```shell
C:\Users\lived>echo %PATH%
C:\Python311\Scripts\;C:\Python311\;<a lot more entries separated by ;>
```

pip complaining about missing SSL support is usually because it can't find the dll for SSL.

**Comment by devdave** [comment]
@donhuvy Since your problem is with how python is set up on your computer, I recommend reddit.com/r/learnpython/ or perhaps better https://discuss.python.org/ as places to seek out help with the interpreter.

**Comment by AlphaeusNg** [comment]
> Currently pytorch with cuda and cp311 can be installed on windows! To install cuda enabled torch on cu311 `pip3 install --pre torch --index-url https://download.pytorch.org/whl/nightly/cu118`
> 
> ![image](https://user-images.githubusercontent.com/20739885/218238308-4d5b4afb-88c5-490f-951a-f7b8c591c7f7.png)
> 
> ![image](https://user-images.githubusercontent.com/20739885/218238419-5f703235-102e-4c34-add3-46b91bfe8f8c.png)

Wow this works. Thanks so much dude, you're a life saver!

**Comment by DrSky2007** [comment]
ROCm Support is still missing in python 3.11, even in nightly builds.
With python 3.10 everything works as expected, but when I try to install pytroch with python 3.11 I get this error:
`pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/rocm5.3
Looking in indexes: https://download.pytorch.org/whl/nightly/rocm5.3`
`ERROR: Could not find a version that satisfies the requirement torch (from versions: none)`
`ERROR: No matching distribution found for torch`

**Comment by Alon-Retispec** [comment]
ROCm is not supported on Windows PyTorch at all (for any Python version)

**Comment by DrSky2007** [comment]
Ok, but I am using the linux version...

**Comment by johnnynunez** [comment]
> ROCm is not supported on Windows PyTorch at all (for any Python version) ![image](https://user-images.githubusercontent.com/69560457/221196372-bf370f71-5da3-4435-af29-21c4479b67a3.png)

There are leaks that AMD said on their website that for windows, "coming soon". After a few hours they closed the page.

**Comment by Alon-Retispec** [comment]
@DrSky2007 sorry, I don't know why I thought you were talking about Win.

**Comment by alex77g2** [comment]
> @dagitses suggests that it's probably OK to disable JIT tests and have a followup issue for supporting JIT in 3.11. Also please note that 3.11 is yet not available on Anaconda: https://anaconda.org/anaconda/python

It seems that Anaconda 3.11.0 is now available for all 64-bit OS (except Win32 and linux-32).
https://anaconda.org/anaconda/python
Does solve all open dependencies?

**Comment by johnnynunez** [comment]
> > @dagitses suggests that it's probably OK to disable JIT tests and have a followup issue for supporting JIT in 3.11. Also please note that 3.11 is yet not available on Anaconda: https://anaconda.org/anaconda/python
> 
> It seems that Anaconda 3.11.0 is now available for all 64-bit OS (except Win32 and linux-32). https://anaconda.org/anaconda/python Does solve all open dependencies?

The problem is not python itself, which has been there for a month in anaconda channel stable. It is that the main libraries like numpy, still anaconda has not uploaded the python 3.11 binaries

**Comment by adam2392** [comment]
@johnnynunez just following up on this to try to track what else is needed. Sorry if this is a redundant question, but I see Python 3.11 wheels here: https://anaconda.org/scipy-wheels-nightly/numpy/files

Is there something I should look out for specifically to know when to upgrade our package release (which depends on PyTorch) to Python 3.11?

**Comment by johnnynunez** [comment]
> @johnnynunez just following up on this to try to track what else is needed. Sorry if this is a redundant question, but I see Python 3.11 wheels here: https://anaconda.org/scipy-wheels-nightly/numpy/files
> 
> Is there something I should look out for specifically to know when to upgrade our package release (which depends on PyTorch) to Python 3.11?

https://anaconda.org/anaconda/scipy/files

Not nightly versions.
It supposed that in the following weeks, the will be released python 3.11 binaries

**Comment by johnnynunez** [comment]
> I've build all necessary dependencies for Python-3.11 support in `pytorch-nightly` channel, but indeed as Anaconda does not have support for Python-3.11 in time for 2.0.0 release, those wouldn't be available in `pytorch` conda channels, though wheels should be available for all platforms.

Totally reasonable! Do you know more or less the release date of pytorch 2.0? Will there be update soon with cuda 12.1, after pytorch 2.0 release?

**Comment by Birch-san** [comment]
@johnnynunez pytorch 2.0 expected to release [March 15th](https://github.com/pytorch/pytorch/issues/77764#issuecomment-1461797335).

**Comment by whaaften123** [comment]
So conclusion we canth use pytorch on windows now? Or we can if we downgrade to version 3.10?

**Comment by johnthagen** [comment]
[PyTorch 2.0.0 has been released onto PyPI with 3.11 wheels](https://pypi.org/project/torch/2.0.0/#files).

**Comment by EwoutH** [comment]
Thanks everybody! If no one objects, I will close this issue as completed.

**Comment by eric-tc-wong** [comment]
Was model.compile() supposed to be supported with python 3.11 and pytorch 2.0.0?
I am getting the error
`File /usr/local/lib/python3.11/dist-packages/torch/_dynamo/eval_frame.py:377, in check_if_dynamo_supported()
    375     raise RuntimeError("Windows not yet supported for torch.compile")
    376 if sys.version_info >= (3, 11):
--> 377     raise RuntimeError("Python 3.11+ not yet supported for torch.compile")

RuntimeError: Python 3.11+ not yet supported for torch.compile`

**Comment by johnnynunez** [comment]
Conda binaries was uploaded today

**Comment by ivaquero** [comment]
I have a question: Why python 3.11 is still not supported on conda's `pytorch` channel, since it has been supported on `pytorch-nightly` channel for quite a while ? Is it a issue on anaconda side?

**Comment by saiden89** [comment]
Hey, are the Python 3.11 wheels for ROCm planned? I still cannot find them in the index as they stop at 3.10.

**Comment by DrSky2007** [comment]
> Hey, are the Python 3.11 wheels for ROCm planned? I still cannot find them in the index as they stop at 3.10.

Same issue here!

**Comment by LeenHawk** [comment]
> I have a question: Why python 3.11 is still not supported on conda's `pytorch` channel, since it has been supported on `pytorch-nightly` channel for quite a while ? Is it a issue on anaconda side?

Same issue

**Comment by payne911** [comment]
Running:

> pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu117

Gave me:

>   Downloading https://download.pytorch.org/whl/certifi-2022.12.7-py3-none-any.whl (155 kB)
     ---------------------------------------- 155.3/155.3 kB 3.1 MB/s eta 0:00:00
Requirement already satisfied: mpmath>=0.19 in c:\python311\lib\site-packages (from sympy->torch) (1.2.1)
Installing collected packages: urllib3, pillow, numpy, MarkupSafe, idna, charset-normalizer, certifi, requests, jinja2, torch, torchvision, torchaudio
  WARNING: Failed to write executable - trying to use .deleteme logic
ERROR: Could not install packages due to an OSError: [WinError 2] The system cannot find the file specified: 'C:\\Python311\\Scripts\\f2py.exe' -> 'C:\\Python311\\Scripts\\f2py.exe.deleteme'

**Comment by Bill-Gots** [comment]
> Was torch.compile() supposed to be supported with python 3.11 and pytorch 2.0.0? I am getting the error
> 
> ```
> File /usr/local/lib/python3.11/dist-packages/torch/_dynamo/eval_frame.py:377, in check_if_dynamo_supported()
>     375     raise RuntimeError("Windows not yet supported for torch.compile")
>     376 if sys.version_info >= (3, 11):
> --> 377     raise RuntimeError("Python 3.11+ not yet supported for torch.compile")
> 
> RuntimeError: Python 3.11+ not yet supported for torch.compile
> ```

Same issue here, sadly

**Comment by Bill-Gets** [comment]
> Was torch.compile() supposed to be supported with python 3.11 and pytorch 2.0.0? I am getting the error
> 
> ```
> File /usr/local/lib/python3.11/dist-packages/torch/_dynamo/eval_frame.py:377, in check_if_dynamo_supported()
>     375     raise RuntimeError("Windows not yet supported for torch.compile")
>     376 if sys.version_info >= (3, 11):
> --> 377     raise RuntimeError("Python 3.11+ not yet supported for torch.compile")
> 
> RuntimeError: Python 3.11+ not yet supported for torch.compile
> ```

I met the same error here in `eval_frame.py.`. Although I can use pytorch 2.0 with python 3.11, it seems that I can't use `torch.compile` because of this issue.

**Comment by malfet** [comment]
> Running:
> 
> > pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu117
> 
> Gave me:
> 
> > Downloading https://download.pytorch.org/whl/certifi-2022.12.7-py3-none-any.whl (155 kB)
> > ---------------------------------------- 155.3/155.3 kB 3.1 MB/s eta 0:00:00
> > Requirement already satisfied: mpmath>=0.19 in c:\python311\lib\site-packages (from sympy->torch) (1.2.1)
> > Installing collected packages: urllib3, pillow, numpy, MarkupSafe, idna, charset-normalizer, certifi, requests, jinja2, torch, torchvision, torchaudio
> > WARNING: Failed to write executable - trying to use .deleteme logic
> > ERROR: Could not install packages due to an OSError: [WinError 2] The system cannot find the file specified: 'C:\Python311\Scripts\f2py.exe' -> 'C:\Python311\Scripts\f2py.exe.deleteme'

@payne911 can you please file a separate issue, and share a bit more about your system details, as I can not reproduce it.

**Comment by sweettyler** [comment]
"RuntimeError: Python 3.11+ not yet supported for torch.compile"

When will this be solved?

**Comment by alex77g2** [comment]
Could someone sum up the status please?
I can install (without error) pytorch on Python 3.11.

Microsoft Windows [Version 10.0.19045.2846]
Python 3.11.3
pip 23.1 from C:\...\AppData\Roaming\Python\Python311\site-packages\pip (python 3.11)
pip3 install torch torchvision torchaudio # today, not using --index-url
#pip list (cropped)
numpy              1.24.2
torch              2.0.0
torchaudio         2.0.1
torchvision        0.15.1

**Comment by LeenHawk** [comment]
> Could someone sum up the status please? I can install (without error) pytorch on Python 3.11.
> 
> Microsoft Windows [Version 10.0.19045.2846] Python 3.11.3 pip 23.1 from C:...\AppData\Roaming\Python\Python311\site-packages\pip (python 3.11) pip3 install torch torchvision torchaudio # today, not using --index-url #pip list (cropped) numpy 1.24.2 torch 2.0.0 torchaudio 2.0.1 torchvision 0.15.1

Do you try torch.compile?

**Comment by LeenHawk** [comment]
> > 
> 
> No, I did not try to compile. But perhaps I get the original issue wrong. I thought it is about Pytorch normal usage, which I thought does not require to compile it. Then I would like to raise my question: can I use Pytorch with Python311 under Windows and Linux, or is compiling part of typical usage?

Emm, torch.compile is a main feature of pytorch 2.0 in my view, just for me.

**Comment by adaaaaaa** [comment]
> raise RuntimeError("Python 3.11+ not yet supported for torch.compile")

same issue...

**Comment by johnnynunez** [comment]
> > raise RuntimeError("Python 3.11+ not yet supported for torch.compile")
> 
> same issue...

https://github.com/pytorch/pytorch/pull/99180#issuecomment-1535995426 they are working on that

**Comment by MikulasZelinka** [comment]
https://github.com/pytorch/pytorch/issues/86566 is now merged, but I'm not sure when this will be released.

**Comment by MikulasZelinka** [comment]
@malfet @albanD Looks like all tasks are done, do you know when this will be released please (including support for `torch.compile`?

**Comment by malfet** [comment]
@MikulasZelinka sometime in Oct, see https://dev-discuss.pytorch.org/t/pytorch-release-2-1-0/1271

In the meantime, all features one wants should be available in nightly builds.

**Comment by malfet** [comment]
Closing the issue, as 3.11 support (except for compile) is available as of PyTorch-2.0 release

**Comment by garfieldnate** [comment]
Except for torch.compile, I guess. I came here because I received this error:

    Python 3.11+ not yet supported for torch.compile

**Comment by ancestor-mithril** [comment]
> torch.compile, I guess. I came here because I received

Pytorch 2.0.1 should support both Python 3.11 and torch.compile.

**Comment by garfieldnate** [comment]
@ancestor-mithril Interesting, maybe the release got messed up or something? I'm using torch 2.0.1, python 3.11.2 on an M1 Mac. `torch.version.__version__` gives `2.0.1` and `torch.version.git_version` gives `e9ebda29d87ce0916ab08c06ab26fd3766a870e5`.  You can see [here](https://github.com/pytorch/pytorch/blob/e9ebda29d87ce0916ab08c06ab26fd3766a870e5/torch/_dynamo/eval_frame.py#L376) that the old version check is still there in this commit.

If you download the wheel from PyPi (I'm using [OSX- arm64](https://files.pythonhosted.org/packages/85/68/f901437d3e3ef6fe97adb1f372479626d994185b8fa06803f5bdf3bb90fd/torch-2.0.1-cp311-none-macosx_11_0_arm64.whl)) and inspect the source there directly, you'll see the old check in place there, too.

Should I open a new ticket for this? There may be other things missing from the release, as well.

***

Here's my env info:

```
PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: version 3.25.1
Libc version: N/A

Python version: 3.11.2 (main, Feb 16 2023, 02:55:59) [Clang 14.0.0 (clang-1400.0.29.202)] (64-bit runtime)
Python platform: macOS-13.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Max

Versions of relevant libraries:
[pip3] numpy==1.25.0
[pip3] torch==2.0.1
[pip3] torchviz==0.0.2
[conda] Could not collect
```

Trying to use torch.compile leads to this error:

```
    model = torch.compile(model)
            ^^^^^^^^^^^^^^^^^^^^
  File ".../venv/lib/python3.11/site-packages/torch/__init__.py", line 1441, in compile
    return torch._dynamo.optimize(backend=backend, nopython=fullgraph, dynamic=dynamic, disable=disable)(model)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../venv/lib/python3.11/site-packages/torch/_dynamo/eval_frame.py", line 413, in optimize
    check_if_dynamo_supported()
  File "...venv/lib/python3.11/site-packages/torch/_dynamo/eval_frame.py", line 377, in check_if_dynamo_supported
    raise RuntimeError("Python 3.11+ not yet supported for torch.compile")
RuntimeError: Python 3.11+ not yet supported for torch.compile
```

**Comment by mlazos** [comment]
cc @williamwen42

**Comment by malfet** [comment]
Pytorch-2.0.1 does not support `torch.compile` yet, but it should be available in the nightly builds.

**Comment by d-kleine** [comment]
@malfet Just tested the most recent nightly build (`2.1.0.dev20230703`), both for cuda 11.8 and 12.1: `torch.compile()` still does not work (yet)

> RuntimeError: Windows not yet supported for torch.compile

**Comment by brianmatzelle** [comment]
Bypass this on Windows by adding `--compile=False` to the end of your run statement. Ex:

`python train.py config/finetune_conversation.py --compile=False` works, whereas
`python train.py config/finetune_conversation.py` will raise the error

**Comment by malfet** [comment]
@d-kleine yes, neither cpp compile backend nor triton currently work on Windows, though this is not specific to 3.11, so feel free to open a separate issue (unless one already exits, if that's the case, please ping me on it)
cc: @vladimir-aubrecht

**Comment by d-kleine** [comment]
@malfet Should be covered by #90768, what do you think?

**Comment by georgeseifada** [comment]
Will python 3.11 be supported for torch `1.13.1`? I get an error from poetry when trying to install it:

```
RuntimeError

  Unable to find installation candidates for torch (1.13.1)

  at ~/.virtualenvs/sentence-embedding/lib/python3.11/site-packages/poetry/installation/chooser.py:76 in choose_for
       72│ 
       73│             links.append(link)
       74│ 
       75│         if not links:
    →  76│             raise RuntimeError(f"Unable to find installation candidates for {package}")
       77│ 
       78│         # Get the best link
       79│         chosen = max(links, key=lambda link: self._sort_key(package, link))
       80│ 

```

**Comment by DarkAlchy** [comment]
> Will python 3.11 be supported for torch `1.13.1`? I get an error from poetry when trying to install it:
> 
> ```
> RuntimeError
> 
>   Unable to find installation candidates for torch (1.13.1)
> 
>   at ~/.virtualenvs/sentence-embedding/lib/python3.11/site-packages/poetry/installation/chooser.py:76 in choose_for
>        72│ 
>        73│             links.append(link)
>        74│ 
>        75│         if not links:
>     →  76│             raise RuntimeError(f"Unable to find installation candidates for {package}")
>        77│ 
>        78│         # Get the best link
>        79│         chosen = max(links, key=lambda link: self._sort_key(package, link))
>        80│ 
> ```

Probably about the time 3.12 comes out. /shrug.

**Comment by sytelus** [comment]
It's been almost a year Python 3.11 was released and PyTorch still doesn't support it in stable release :(. Any idea, when the next stable release with Python 3.11 support will be available?

**Comment by AntreasAntoniou** [comment]
I am wondering that too, since I consider Pytorch's stable releases to be a heuristic for which python version I should be using. :D

But other than that, for all the sweet and promised gains.

**Comment by johnnynunez** [comment]
> It's been almost a year Python 3.11 was released and PyTorch still doesn't support it in stable release :(. Any idea, when the next stable release with Python 3.11 support will be available?


pytorch 2.1.0 rc1 is coming.  See tracker in issues. Release on november- end of the year

**Comment by crypdick** [comment]
Just in time for things to break with the new Python 3.12 release ;) https://github.com/pytorch/pytorch/issues/114893
